docs/lang/articles/faqs/faq.md

```diff
@@ -66,3 +66,112 @@ maintenance overheads.
+
+If you use the default `ti.f32`, then this is normal because `float32` has only six to seven decimal digits of precision. Even a constant is cast to a `float32` number by default. To solve this issue, declare `default_fp=ti.f64` in `ti.init`.
+
+For example, if you stay with `float32`:
+
+```python
+import taichi as ti
+ti.init(debug=True)
+@ti.func
+def get_f_stress3(f_stress):
+    res = ti.Matrix([[f_stress[0], f_stress[2], 0.0], [f_stress[2], f_stress[1], 0.0], [0.0, 0.0, f_stress[3]]], ti.f64)
+    return res
+@ti.kernel
+def foo():
+    stress = get_f_stress3(f_stress)
+    print("σ =", stress)
+if __name__ == "__main__":
+    f_stress = ti.Vector([-0.742128, -1.731632, -0.000001, -0.742128], ti.f64)
+    foo()
+```
+
+The output becomes:
+
+ ```plaintext
+ [Taichi] version 1.0.0, llvm 10.0.0, commit 6a15da85, win, python 3.8.1
+[Taichi] Starting on arch=x64
+σ = [[-0.742128014565, -0.000001000000, 0.000000000000], [-0.000001000000, -1.731631994247, 0.000000000000], [0.000000000000, 0.000000000000, -0.742128014565]]
+ ```
+
+After switching to `float64` by setting `ti.init(debug=True, default_fp=ti.f64)`, you get the correct output:
+
+```plaintext
+σ = [[-0.742128000000, -0.000001000000, 0.000000000000], [-0.000001000000, -1.731632000000, 0.000000000000], [0.000000000000, 0.000000000000, -0.742128000000]]
+ ```
+
+### Could you recommend some reading materials for cloth simulation, or about multigrid methods in particular?
+
+[This paper](https://www.cs.cmu.edu/~baraff/papers/sig98.pdf) proposes an implicit integration method that enforces constraints on individual cloth particles.
+
+[This paper](https://www.cs.ubc.ca/~rbridson/docs/cloth2002.pdf) presents a collision handling algorithm compatible with any technique that simulates the internal dynamics of a piece of cloth.
+
+As for multigrid methods, you may be interested in an efficient multigrid scheme, which simulates high-resolution deformable objects in their full spaces at interactive frame rates, as explained in [this paper](http://tiantianliu.cn/papers/xian2019multigrid/xian2019multigrid.html).
+
+### Why does it always return an error when I pass a list from the Python scope to a Taichi kernel?
+
+A Taichi kernel cannot take a Python list directly. You need to use NumPy Arrays as a bridge.
+
+For example, the following code snippet does not work:
+
+```python
+import taichi as ti
+ti.init()
+def index_of(a, list):
+    for i in range(0, len(list)):
+        if list[i] == a: return i
+    return -1
+@ti.kernel
+def sort_by_values(list1: list, values: list):
+    sorted_list = []
+    while len(sorted_list) != len(list1):
+        if index_of(min(values), values) in list1:
+            sorted_list.append(index_of(min(values), values))
+        values[index_of(min(values), values)] = math.inf
+    return sorted_list
+l1 = [0, 1, 2, 3, 4, 5]
+vs = [225, 114, 123, 53, 654, 15]
+r = sort_by_values(l1, vs)
```

Comment:
```suggestion
```python
import taichi as ti
import numpy as np
ti.init()
x = ti.field(ti.i32, shape=3)
array = [10, 20, 30]

@ti.kernel
def test(arr: list):
    for i in range(3):
        x[i] = arr[i]
test(array)
```

Reply:
The example is too complex. We need a simple one to make the idea clear.